example_test.go

```diff
@@ -229,3 +229,120 @@ func ExampleDigestSigner() {
 	// Output:
 	// digest signed
 }
+
+// This example demonstrates signing and verifying countersignatures.
+//
+// The COSE Countersignature API is EXPERIMENTAL and may be changed or removed in a later
+// release.
+func ExampleCountersignature() {
+	// create a signature holder
+	sigHolder := cose.NewSignature()
+	sigHolder.Headers.Protected.SetAlgorithm(cose.AlgorithmES512)
+	sigHolder.Headers.Unprotected[cose.HeaderLabelKeyID] = []byte("1")
+
+	// create message to be signed
+	msgToSign := cose.NewSignMessage()
+	msgToSign.Payload = []byte("hello world")
+	msgToSign.Signatures = append(msgToSign.Signatures, sigHolder)
+
+	// create a signer
+	privateKey, _ := ecdsa.GenerateKey(elliptic.P521(), rand.Reader)
+	signer, _ := cose.NewSigner(cose.AlgorithmES512, privateKey)
+
+	// sign message
+	msgToSign.Sign(rand.Reader, nil, signer)
+
+	// create a countersignature holder for the message
+	msgCountersig := cose.NewCountersignature()
+	msgCountersig.Headers.Protected.SetAlgorithm(cose.AlgorithmES512)
+	msgCountersig.Headers.Unprotected[cose.HeaderLabelKeyID] = []byte("11")
+
+	// create a countersigner
+	counterPrivateKey, _ := ecdsa.GenerateKey(elliptic.P521(), rand.Reader)
+	countersigner, _ := cose.NewSigner(cose.AlgorithmES512, counterPrivateKey)
+
+	// countersign message
+	err := msgCountersig.Sign(rand.Reader, countersigner, msgToSign, nil)
+	if err != nil {
+		panic(err)
+	}
+
+	// add countersignature as message unprotected header; notice the
+	// countersignature should be assigned as value not reference
```

Comment:
Why can't we assign it by reference?

Reply:
Good question. If you do then you end up obtaining the wrong marshaled message.

In order to be flexible and accept references, then `MarshalCBOR` function for headers would need to detect countersignature references and replace them by values in runtime, which would add a processing penalty and make the code a little more convoluted.

But that may be done anyways if there is any intention to keep adding other objects as header values moving forward.

A last aspect is that the library would enforce using references instead of values if that's the intended design decision, at unmarshaled structs, so that users can simply detect if the countersignature header refers to a single countersignature reference or a slice of countersignature references. Because of the current CBOR limitations I've opted by value in both cases instead.

Reply:
Eventually I found a way to make it without too many collateral effects.

Reply:
Neat!